### ISSUE TITLE
[Bug] Plain legend label overflows container if text is too long

### ISSUE THREAD

**Issue by mrgawrys** [apache/echarts#20237]
### Version

5.5.1

### Link to Minimal Reproduction

https://echarts.apache.org/examples/en/editor.html?c=bar-y-category-stack&code=PYBwLglsB2AEC8sDeAoWszGAG0iAXMmuhgE4QDmFApqYQOQCGAHhAM70A0x6L7ACsAjQwtQqhIkA9FNgBVNtVh82GYGUo1SanHh4kwATxDUGbABaMAJsADu9WDNj0L1uw8aqr1AGaMArrgA3LAAxoxwjNhs6gBGSvTYwtQOwNouljb2-gC-xDnc6NjUNNBW4vpGJgwg2IzC9PmFsBTk5USSxT5gDADMAKRc-uQU5j3OACyDzeixwGCYALZ90_qhMGD10AAyjPHYhGCk_tRNxMwAgqxsFZJVps4AblEnjegFxIZX7LcGxg_0cKiChpQxDSRWRibQgAbXoAFkYFxnAAVV6cZwAdWoVmR9BR5n8eIAYuQ8QBlKEU_zQegAXTO6EU5GoN1gMP0EkksGgjEWAIAIhBSNRRWLxaLQmBwdz7gxYoxSDLJGxNqEANYMTCbbDKkh1fa_bmwCx2Q7HU7Gj7G6iLECWNg_DrG2A-YChfxsly0CCst7c63cyHQ9m9ABMAAYMb0I2HoxGAIzR3oTaMATijsF6MejkYZkkD6C5kl5_IY8Pq2Dq1aiUVg2CsDZx3issBbzaUsAuuJmf2qzgVSt7TLVmuc2qieqKe2oB2dxtNtnNJ3072HsFt9s8TuL3LdHq9zN9HFXsELJGDjFhCcjGIT4bvibvKYxGYxYZzsDDCYj-ZI593dBSwBC4fB8CAkihZtWwuHE4PbHEADo23gjBoK7eDMKQlCWzQqx8Iw9sEKsZCiKnDB_nlRVyNVRgNS1eZJ3XA1ZyNblF2XS0A3XTcHR3U90H3T0zB9P1T3PdBL1hMNb1gBMAA44zktMky_F8vzfLNP16H8_zXTlT2AhgADUIG8dRu3IuUB2o9daPo8dGN1ZiZznQCSA4sgVytHi7T4tl3ME91hOcI8xJ808pPZBMAFZM2_JTI1U2LU2UzNs0zCZdNyXt3KM5xyWoRVQnMWAAFFoAoCBCuK0qKqqmrSBK8rKuERrmvqtqiqaurWugaB2t6qqBu6jq-sGlrhomzqRtqyaurmmbpvG0ahoWnr5tmjaZugFJ12s-hBxo0cGJ1ciWLcgSTXMM0vK4gtfK3R0AquoTD1Ek8IuNKKYUUzN5IfWAM1UtN1JvTT70_e881yYgGRyIIgA

### Steps to Reproduce

1. Create any chart with multiple series
2. Add legend of type 'plain'
3. Set very long name for one of the series

### Current Behavior

Legend label overflows the screen

### Expected Behavior

Legend label has a break somewhere in the text and continues on the next line 
or
Legend label is truncated with ellipsis

**Comment by helgasoft** [comment]
```
  legend: {
    type: 'plain', textStyle:{width:200, overflow:'truncate'}
  }
```
📌 please close issue if problem solved.

**Comment by mrgawrys** [comment]
Yes, to an extent it does, but rather as a workaround than a fix. Proposed way requires us to set maximum width at an arbitrary length. For me at least, expected behavior would be to have the possibility for the label to fill all available space and truncate if it's not possible to display it whole.

**Comment by helgasoft** [comment]
> label to fill all available space and truncate if it's not possible to display it whole.

Makes sense. Tried to do it with _rich_ but found several _width_ problems:
```
  legend: {
    type: 'plain', 
    width: 22,    // does not work
    formatter: '{trr|{name}}',
    textStyle:{ 
      width: 100,  // does not work along with formatter
      overflow: 'truncate',
      rich: {
        trr: { 
          width: '33%',   // does not work
          color: 'violet'
        }
      }
    },
  },
```

**Comment by helgasoft** [comment]
👍 for digging deeper. ```LegendView._createItem‎()``` is similar to this:
https://github.com/apache/echarts/blob/9a2018537e25c2f3d4250947cc324ff253bfb878/src/component/visualMap/PiecewiseView.ts#L75-L88
which is part of a recent fix to a similar problem (#20956) by @plainheart 
Just trying to connect the dots...

**Comment by ddkwork** [comment]
> The problem is visible on many scenarios where the legend container width is defined and the legend items are long. Here are some hints for whom may want to address the root cause:
> 
> The issue seems to be in `‎LegendView._createItem‎()`, it's rendering the legend item text at the full width, regardless of the legend container width or the legend icon+spacing;
> 
> [echarts/src/component/legend/LegendView.ts](https://github.com/apache/echarts/blob/ef926fa61d41588a663717c4fd1c0412c57474db/src/component/legend/LegendView.ts#L465-L474)
> 
> Lines 465 to 474 in [ef926fa](/apache/echarts/commit/ef926fa61d41588a663717c4fd1c0412c57474db)
> 
>  itemGroup.add(new graphic.Text({ 
>      style: createTextStyle(textStyleModel, { 
>          text: content, 
>          x: textX, 
>          y: itemHeight / 2, 
>          fill: textColor, 
>          align: textAlign, 
>          verticalAlign: 'middle' 
>      }, {inheritColor: textColor}) 
>  })); 
> It should have a defined max-width passed down from `render()` to the `_createItem()` method, here's a partial patch applied directly on the built package:
> 
> diff --git a/node_modules/echarts/lib/component/legend/LegendView.js b/node_modules/echarts/lib/component/legend/LegendView.js
> index ffaaa42..5070daf 100644
> --- a/node_modules/echarts/lib/component/legend/LegendView.js
> +++ b/node_modules/echarts/lib/component/legend/LegendView.js
> @@ -102,7 +102,6 @@ var LegendView = /** @class */function (_super) {
>      if (selector && (!selectorPosition || selectorPosition === 'auto')) {
>        selectorPosition = orient === 'horizontal' ? 'end' : 'start';
>      }
> -    this.renderInner(itemAlign, legendModel, ecModel, api, selector, orient, selectorPosition);
>      // Perform layout.
>      var positionInfo = legendModel.getBoxLayoutParams();
>      var viewportSize = {
> @@ -111,6 +110,7 @@ var LegendView = /** @class */function (_super) {
>      };
>      var padding = legendModel.get('padding');
>      var maxSize = layoutUtil.getLayoutRect(positionInfo, viewportSize, padding);
> +    this.renderInner(itemAlign, legendModel, ecModel, api, selector, orient, selectorPosition, maxSize);
>      var mainRect = this.layoutInner(legendModel, itemAlign, maxSize, isFirstRender, selector, selectorPosition);
>      // Place mainGroup, based on the calculated `mainRect`.
>      var layoutRect = layoutUtil.getLayoutRect(zrUtil.defaults({
> @@ -255,7 +255,7 @@ var LegendView = /** @class */function (_super) {
>        enableHoverEmphasis(labelText);
>      });
>    };
> -  LegendView.prototype._createItem = function (seriesModel, name, dataIndex, legendItemModel, legendModel, itemAlign, lineVisualStyle, itemVisualStyle, legendIcon, selectMode, api) {
> +  LegendView.prototype._createItem = function (seriesModel, name, dataIndex, legendItemModel, legendModel, itemAlign, lineVisualStyle, itemVisualStyle, legendIcon, selectMode, api, maxSize) {
>      var drawType = seriesModel.visualDrawType;
>      var itemWidth = legendModel.get('itemWidth');
>      var itemHeight = legendModel.get('itemHeight');
> @@ -308,7 +308,8 @@ var LegendView = /** @class */function (_super) {
>          y: itemHeight / 2,
>          fill: textColor,
>          align: textAlign,
> -        verticalAlign: 'middle'
> +        verticalAlign: 'middle',
> +        width: maxSize.width - textX,
>        }, {
>          inheritColor: textColor
>        })
> And then there's also `hitRect`, which needs to be dealt with, I've just patched it away for testing purposes.
> 
> And the same is probably true for the scrollable legend view.

hello you, are you planning to implement such a widget at a later stage?
https://github.com/ddkwork/ux

I found the interaction behavior of the right click context menu strange during the implementation of this widget, so I'm looking forward to your implementation of a right click context menu in your way, thanks.
